custom_components/rpi_gpio/cover.py

```diff
@@ -50,88 +53,99 @@
-                cover.get(CONF_UNIQUE_ID),
+                "AS_IS",
+                "PUSH_PULL",
+                cover.get(CONF_STATE_PIN),
+                state_pull_mode,
+                invert_state,
+                cover.get(CONF_UNIQUE_ID) or f"{DOMAIN}_{cover.get(CONF_RELAY_PORT) or cover.get("relay_pin")}_{cover[CONF_NAME].lower().replace(' ', '_')}",
             )
         )
-    add_entities(covers)
-
-
-class RPiGPIOCover(CoverEntity):
-    """Representation of a Raspberry GPIO cover."""
-
-    def __init__(
-        self,
-        name,
-        relay_pin,
-        state_pin,
-        state_pull_mode,
-        relay_time,
-        invert_state,
-        invert_relay,
-        unique_id,
-    ):
-        """Initialize the cover."""
+
+    async_add_entities(covers)
+
+class GPIODCover(CoverEntity):
+    _attr_should_poll = False
+
+    def __init__(self, hub, name, relay_port, relay_time, relay_active_low, relay_bias, relay_drive,
+                 state_port, state_bias, state_active_low, unique_id):
+        _LOGGER.debug(f"GPIODCover init: {relay_port}:{state_port} - {name} - {unique_id} - {relay_time}")
+        self._hub = hub
         self._attr_name = name
         self._attr_unique_id = unique_id
-        self._state = False
-        self._relay_pin = relay_pin
-        self._state_pin = state_pin
-        self._state_pull_mode = state_pull_mode
+        self._relay_port = relay_port
         self._relay_time = relay_time
-        self._invert_state = invert_state
-        self._invert_relay = invert_relay
-        setup_output(self._relay_pin)
-        setup_input(self._state_pin, self._state_pull_mode)
-        write_output(self._relay_pin, 0 if self._invert_relay else 1)
+        self._relay_active_low = relay_active_low
+        self._relay_bias = relay_bias
+        self._relay_drive = relay_drive
+        self._state_port = state_port
+        self._state_bias = state_bias
+        self._start_active_low = state_active_low
+        self._attr_is_closed = False != state_active_low
+        hub.add_cover(self, relay_port, relay_active_low, relay_bias, relay_drive,
+                      state_port, state_bias, state_active_low)
 
     def update(self):
-        """Update the state of the cover."""
-        self._state = read_input(self._state_pin)
-
-    @property
-    def is_closed(self):
-        """Return true if cover is closed."""
-        return self._state != self._invert_state
-
-    def _trigger(self):
-        """Trigger the cover."""
-        write_output(self._relay_pin, 1 if self._invert_relay else 0)
-        sleep(self._relay_time)
-        write_output(self._relay_pin, 0 if self._invert_relay else 1)
+        self._attr_is_closed = self._hub.update(self._state_port)
+        self.schedule_update_ha_state(False)
 
     def close_cover(self, **kwargs):
-        """Close the cover."""
-        if not self.is_closed:
-            self._trigger()
+        if self._attr_is_closed:
```

Comment:
```suggestion
        if self.is_closed:
```
I know this is confusing, but writing should use the `_attr_` and read direct from the property. This allow HA to cache values.

Reply:
You always learn something new. Where do you learn that from?

Reply:
fixed.